File: app/src/main/java/dev/bee/kanjianki/MainActivitySettingsUpdateFlow.java
package dev.bee.kanjianki;

import android.widget.Toast;

import dev.bee.kanjianki.update.GitHubUpdater;
import dev.bee.kanjianki.updatecore.UpdateRunScreenCopy;

final class MainActivitySettingsUpdateFlow {
    private final MainActivitySettings activity;

    MainActivitySettingsUpdateFlow(MainActivitySettings activity) {
        this.activity = activity;
    }

    void runUpdate(boolean cachedPending) {
        activity.base(activity.NAV_SETTINGS_ROUTE);
        int updateUiRun = ++activity.updateUiRunCounter;
        activity.activeUpdateUiRunToken = updateUiRun;
        UpdateRunScreenCopy.Copy copy = UpdateRunScreenCopy.forRun(cachedPending);
        activity.content.addView(MainActivitySettingsUpdateRunCompose.settingsUpdateRunView(
                activity,
                copy.title(),
                copy.body(),
                copy.progressLabel()
        ));
        activity.io.execute(() -> {
            GitHubUpdater updater = new GitHubUpdater(activity);
            GitHubUpdater.UpdateResult result = cachedPending
                    ? updater.installCachedPendingUpdate(GitHubUpdater.UpdateSource.CACHED)
                    : updater.checkDownloadAndInstall(GitHubUpdater.UpdateSource.MANUAL);
            activity.main.post(() -> {
                if (activity.activeUpdateUiRunToken != updateUiRun) {
                    return;
                }
                Toast.makeText(activity, result.message, Toast.LENGTH_LONG).show();
                if (result.intent != null) {
                    activity.startActivity(result.intent);
                }
                activity.renderUpdate();
            });
        });
    }
}

File: app/src/main/kotlin/dev/bee/kanjianki/MainActivitySettingsUpdateFlow.kt
package dev.bee.kanjianki

import android.widget.Toast
import dev.bee.kanjianki.update.GitHubUpdater
import dev.bee.kanjianki.updatecore.UpdateRunScreenCopy

internal class MainActivitySettingsUpdateFlow(private val activity: MainActivitySettings) {
    fun runUpdate(cachedPending: Boolean) {
        val copy = UpdateRunScreenCopy.forRun(cachedPending)
        activity.composeRoute(MainActivityBase.NAV_SETTINGS_ROUTE) {
            SettingsUpdateRunScreen(
                model = SettingsUpdateRunModel(
                    title = copy.title(),
                    body = copy.body(),
                    progressLabel = copy.progressLabel(),
                    onHome = activity::renderHome,
                    onBack = { activity.renderSettings(false) },
                )
            )
        }
        val updateUiRun = ++activity.updateUiRunCounter
        activity.activeUpdateUiRunToken = updateUiRun
        activity.io.execute {
            val updater = GitHubUpdater(activity)
            val result = if (cachedPending) {
                updater.installCachedPendingUpdate(GitHubUpdater.UpdateSource.CACHED)
            } else {
                updater.checkDownloadAndInstall(GitHubUpdater.UpdateSource.MANUAL)
            }
            activity.main.post {
                if (activity.activeUpdateUiRunToken != updateUiRun) {
                    return@post
                }
                Toast.makeText(activity, result.message, Toast.LENGTH_LONG).show()
                result.intent?.let(activity::startActivity)
                activity.renderUpdate()
            }
        }
    }
}
